### PR TITLE
Lighting a fire is hindered by wind and rain

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "Lighting a fire is hindered by the elements"```

#### Purpose of change
As part of the continuing wind series... this will make lighting a fire a more difficult operation in the wind and rain.
#### Describe the solution
Lighting a fire, can be a long-activity , if using a magnifying glass or fire drill.
When using a lighter, the time-taken is usually less than 2 turns so no long activity.
Lighting an entire tree on fire, in the wind and rain, with a disposable lighter, in 6 seconds, does not seem realistic to me.
So I've added a time multiplier to the activity, dependent on the wind and precipitation.
This abstracts the time it would take a survivor to find dry bits, fail to catch a spark, etc.
In high winds and rain, lighting a fire succesfully may take around 30 minutes now.

#### Describe alternatives you've considered
Further work in this area is encouraged, firestarting is quite basic, lighting a tree should not be the same as lighting a shrub, or a two-by-four, or even the wall of a house, yet all are similar now.

#### Additional context
N/A

### CODE REVIEW

**Comment by I-am-Erk** [comment]
In discord you've talked about how a big step to this will be getting items to track their wetness.

Besides that I'd say that firestarting depends on a couple things:

- surface area to weight ratio of the item you're starting. Paper and tinder light faster than a log. This could be abstracted as weight, perhaps? Or for maximum granularity, volume and weight where higher volume increases flammability and higher weight reduces it, but volume drops off geometrically while weight is linear (if I'm remembering my curves right)

- heat of the firestarting tool. This could be represented by a firestarting quality level. An acetylene torch lights fires in larger things faster than a lighter or a match does. Some lighters burn hotter than others. Stuff like that.

**Comment by davidpwbrown** [comment]
firestarting quality level would be easy.
Surface area/density of object bit harder.
Terrain/furniture tracking of how difficult to set those on fire = difficult to not just hardcode everything or just add 10 more flags to JSON.
What we really need is per-terrain data over time - i.e its been raining for 6 hours , this terrain tile now has WET level 6.
I have no idea how to approach or even if it would be feasible performance -wise to track all those tiles in the reality bubble.

**Comment by Vasyan2006** [comment]
There is another unfinished firestarting PR #23699 you can use.

**Comment by I-am-Erk** [comment]
I would say you could get away with making terrain very hard to ignite, furniture slightly hard, and not worry about the difference between a chair and a bookshelf, or a log versus wood wall. At most two flags, one for flammable and one for poorly flammable.

Likewise I don't think terrain needs as much wetness tracking as items. Don't track per terrain tile, just a function to ask the terrain when it last rained on it, and maybe how rainy it was. In general we shouldn't be lighting terrain and furniture on fire directly, we should be igniting them with larger fires, and that isn't as dependent on weather.

**Comment by Xhuis** [comment]
Make sure that containing furniture ignores stuff like this. Ovens, for instance, contain fires if you light them inside, so it would be silly if rain would make it harder to light a fire in an oven when the oven fully contains that fire.

**Comment by davidpwbrown** [comment]
Well, it ignores any difficulty if the fire lighting is done inside or done whilst sheltered.
If you drag an oven outside, and try and light a fire inside it, whilst you are outside in the wind and rain, it would still be difficult.

**Comment by NickGrease** [comment]
Just adding one word to the conversation in case it was overlooked or to decide if it's even applicable to this PR:

"Accelerant"

**Comment by davidpwbrown** [comment]
The next step would be to look at that previous firestarting PR and incorporate things like dry tinder, accelerant and other tools/techniques to improve firestarting.

**Comment by Xhuis** [comment]
Dry tinder does actually exist already, so it'd be a nice next step!

**Comment by davidpwbrown** [comment]
Tinder exists in more survival tools mod, but only as a fuel source for the hobo stove, nothing else .

**Comment by Xhuis** [comment]
Supposedly if accelerants were added, it could be flagged as one. I'm not saying it's necessary, just giving my `cash card ($0.02)`.

**Comment by davidpwbrown** [comment]
I wanted to mainline more survival tools, for that reason, the ember carrier also. Kevin suggested that mod would need a lot of balancing to mainline it, so I may just pick the firestarting bits from it.

**Comment by kevingranade** [comment]
I agree it doesnt seem necessary to track precipitation on a per-tile basis, tracking it per-submap should be fine.  Since items can move around they need to carry their state with them though.

**Comment by davidpwbrown** [comment]
I'm gonna mark this as WIP for the moment, I feel it needs more things to mitigate the increased difficulty and more in-depth mechanics before just dumping the increased difficulty on people. I have some ideas though.

**Comment by davidpwbrown** [comment]
Ok so ive finished this up.
I've mainlined tinder and hobo stove from more survival tools.
Tinder is now also a fire accelerant, along with fuels ( diesel, napalm etc ) 
I've added firestarting quality levels to various tools.
And they all have base use move times aswell.
Flint & steel, fire drills, matchbox = level 1 ( but matchbox is quicker )
lighter and electric firestarter = level 2
refillable lighter and flaming weapons = level 3
WHen using a firestarting item, the game will check the last 24 hours for accumulated rainfall, this will determine how "wet " the ground and item being burnt is as an abstraction of a buildup of groundwater, wet branches etc.
( this only happens when outside - indoors firestarting is easy ) 
It will also check local wind ( including if a nearby wind blocking terrain is sheltering )
it will check survival skill, and the items firestarting quality.
and it will calculate the time it may take to start a fire, if that time is over 2 hours, it's outright impossible, and wont start.
This will usually only happen with poor tools, and/or low survival skill and/or poor conditions.
If you have accelerant on you , it will prompt to use it, this will reduce the time taken below that 120 minute threshold in some cases.
Therefore with flint and steel and napalm, you can probably still start a fire in the pouring rain, but otherwise probably not.
I havn't touched how difficult the actual target may be to set ablaze, that will be in a separate PR
invite comments on balance.

File: src/iuse_actor.cpp [hunk 1]
```diff
@@ -1131,28 +1132,105 @@ long firestarter_actor::use( player &p, item &it, bool t, const tripoint &spos )
...
         return 0;
     }
-
     tripoint pos = spos;
-    float light = light_mod( pos );
     if( !prep_firestarter_use( p, pos ) ) {
         return 0;
     }
-
+    int firestart_level = it.get_quality( quality_id( "FIRESTARTING" ) );
```
Comment (Night-Pryanik) [inline src/iuse_actor.cpp]: Could be const.

File: src/iuse_actor.cpp [hunk 2]
```diff
@@ -1131,28 +1132,105 @@ long firestarter_actor::use( player &p, item &it, bool t, const tripoint &spos )
...
+    const time_point update_to = calendar::turn;
+    const time_point update_from = calendar::turn - DAYS( 1 );
+    // get the general wetness of this area from history of rain
+    bool sheltered = g->is_sheltered( pos );
```
Comment (Night-Pryanik) [inline src/iuse_actor.cpp]: Could be const.

File: src/iuse_actor.cpp [hunk 3]
```diff
@@ -1131,28 +1132,105 @@ long firestarter_actor::use( player &p, item &it, bool t, const tripoint &spos )
...
+    auto accum_weather = sum_conditions( update_from, update_to, pos );
+    int wetness = 1;
+    if( !sheltered ) {
+        wetness =  std::max( 1, static_cast<int>( accum_weather.rain_amount / 7000 ) );
+    }
+    const oter_id &cur_om_ter = overmap_buffer.ter( g->m.getabs( pos ) );
+    const w_point weatherPoint = *g->weather_precise;
+    double windpower = weatherPoint.windpower;
+    windpower = get_local_windpower( windpower, cur_om_ter, pos,
```
Comment (Night-Pryanik) [inline src/iuse_actor.cpp]: You declare `windopower` one line above, assign some value to it, and then assign some other value on the very next line.

File: src/iuse_actor.cpp [hunk 4]
```diff
@@ -1131,28 +1132,105 @@ long firestarter_actor::use( player &p, item &it, bool t, const tripoint &spos )
...
+                                     weatherPoint.winddirection, sheltered );
+    float light = light_mod( pos );
```
Comment (Night-Pryanik) [inline src/iuse_actor.cpp]: Could be const.

File: src/iuse_actor.cpp [hunk 5]
```diff
@@ -1131,28 +1132,105 @@ long firestarter_actor::use( player &p, item &it, bool t, const tripoint &spos )
...
     double skill_level = p.get_skill_level( skill_survival );
     /** @EFFECT_SURVIVAL speeds up fire starting */
     float moves_modifier = std::pow( 0.8, std::min( 5.0, skill_level ) );
-    const int moves_base = moves_cost_by_fuel( pos );
+    int moves_base = moves_cost_by_fuel( pos );
```
Comment (Night-Pryanik) [inline src/iuse_actor.cpp]: Why the change? `moves_base` isn't changed anywhere below.

**Comment by davidpwbrown** [comment]
Marking as WIP, as there is a new wind system using debug overrides and fixing various things, that will make this obsolete as it is, itll need to be rebased to #28143

**Comment by Vasyan2006** [comment]
Why refillable lighter is much better than matchbox? They both produce steady fire (but weak and for short time) and should have level 2 quality.
You can add craftable gasoline soaked rag as high quality accelerant.

**Comment by davidpwbrown** [comment]
A refillable lighter ( a zippo ) will resist the wind and rain fine,  using it outdoors is much more resistant to these things.
A match will blow out in high winds very easily, its flame is much more fragile, that is why a refillable lighter is "better".
More charges is a good idea.
Gasoline soaked rag is a good idea.
I will add these after the debug wind is merged and I will then need to rebase these changes anyway.

**Comment by Photoloss** [comment]
Can you take a look at the "mini-flamethrower" CBM too please? Its current stats are in line with the [Looftlighter](https://www.looftlighter.com/en/home/) which would be an exceptionally effective tool for lighting in inclement conditions but the power draw is **way** too high if it's supposed to be a small spark lighter (as in, it draws twice as much energy as a single refillable lighter charge could possibly yield). Also the name is ridiculous no matter how it supposedly functions.

**Comment by RosaEscura** [comment]
If you could also consider fixing rain putting out fires in contained spaces like ovens and wood stoves it'd be much appreciated, here's my issue for it #27128

**Comment by davidpwbrown** [comment]
@RosaEscura  I'll tackle that in another PR

**Comment by davidpwbrown** [comment]
rebased to master, using the wind override system

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
There seems to be some other changes in here that I dont remember doing, perhaps when I rebased , some other things got carried through. 
Ill investigate.

**Comment by halfahermit** [comment]
> Can you take a look at the "mini-flamethrower" CBM too please? Its current stats are in line with the [Looftlighter](https://www.looftlighter.com/en/home/) which would be an exceptionally effective tool for lighting in inclement conditions but the power draw is **way** too high if it's supposed to be a small spark lighter (as in, it draws twice as much energy as a single refillable lighter charge could possibly yield). Also the name is ridiculous no matter how it supposedly functions.

One difference from the lighter you linked is that mini-flamethrower CBM doesn't require any fuel to start a fire. Not sure if that's intentional or not, but it justifies the high power cost somewhat.

**Comment by Photoloss** [comment]
> One difference from the lighter you linked is that mini-flamethrower CBM doesn't require any fuel to start a fire. Not sure if that's intentional or not, but it justifies the high power cost somewhat.

Ugh, did you look at the linked page **at all???** I specifically linked it because it prides itself in **NOT** using any fuel or other chemicals, they [explicitly advertise it as such!](https://www.amazon.com/Looftlighter-70018-Charcoal-Starter/dp/B000WYY65Y) It's basically an electric heating element and a fan, hence the 1500W power draw.

If anything the "mini-flamethrower" could potentially harvest metabolic waste like skin flakes to use as a quasi-free tinder but I wanted to stick to technologies which actually exist IRL.

**Comment by halfahermit** [comment]
Yes, I looked at the page. It shows the lighter being used to set fire to fuel like wood and charcoal. As you said, it's a heating element, not actually capable of projecting fire like a flamethrower or the CBM as it currently works.

**Comment by Photoloss** [comment]
> Yes, I looked at the page. It shows the lighter being used to set fire to fuel like wood and charcoal. As you said, it's a heating element, not actually capable of projecting fire like a flamethrower or the CBM as it currently works.

So, what exactly is the problem? If fires persist on non-flammable surfaces for extended periods of time you should file a bug report as it does not make sense for a regular lighter or a fire drill to provide sufficient fuel either. Out of those the 1500W electric heater will still deliver the *most* heat damage though nothing close to a cooking fire. Hairdryers, teakettles and toasters draw around 1500-2000W, just to give a rough idea of how much heat this thing puts out and how little damage any "lighter" should do without fuel.

**Comment by halfahermit** [comment]
It's not a regular lighter, it's literally called a flamethrower.

**Comment by Photoloss** [comment]
> It's not a regular lighter, it's literally called a flamethrower.

"in name only". Unless I missed a funky update it only lets you light an adjacent tile on fire exactly like a regular match or lighter.

**Comment by halfahermit** [comment]
The difference is matches and lighters need fuel on the target tile, mini-flamethrower does not. If you think it shouldn't work that way, open an issue, I don't want to crowd this PR with it anymore.

**Comment by Photoloss** [comment]
> The difference is matches and lighters need fuel on the target tile, mini-flamethrower does not. If you think it shouldn't work that way, open an issue, I don't want to crowd this PR with it anymore.

That is a problem indeed. And the fire keeps going for 2-3min, no way that can be produced without fuel no matter how much bionic power you pump into it.

**Comment by davidpwbrown** [comment]
I won't look at the mini-flamethrower CBM in this PR.
An in-depth look at CBMs and their power usages is planned however, at some point.

File: src/iuse_actor.cpp [hunk 6]
```diff
@@ -190,7 +191,6 @@ long iuse_transform::use( player &p, item &it, bool t, const tripoint &pos ) con
     if( p.is_worn( *obj ) ) {
         p.reset_encumbrance();
         p.update_bodytemp();
-        p.on_worn_item_transform( *obj );
```
Comment (kevingranade) [inline src/iuse_actor.cpp]: Is this related? I suspect a mismerge.

File: src/iuse_actor.cpp [hunk 7]
```diff
@@ -947,8 +947,7 @@ long deploy_furn_actor::use( player &p, item &it, bool, const tripoint &pos ) co
         return 0;
     }
 
-    // For example: dirt = 2, long grass = 3
```
Comment (kevingranade) [inline src/iuse_actor.cpp]: Likely mismerge.

File: src/iuse_actor.cpp [hunk 8]
```diff
@@ -996,9 +995,6 @@ long reveal_map_actor::use( player &p, item &it, bool, const tripoint & ) const
         p.add_msg_if_player( _( "You should read your %s when you get to the surface." ),
                              it.tname().c_str() );
         return 0;
-    } else if( p.fine_detail_vision_mod() > 4 ) {
```
Comment (kevingranade) [inline src/iuse_actor.cpp]: Mismerge again.

File: src/iuse_actor.cpp [hunk 9]
```diff
@@ -2549,23 +2620,14 @@ bool repair_item_actor::handle_components( player &pl, const item &fix,
                                             ceil( fix.volume() / 250_ml * cost_scaling ) :
                                             divide_roll_remainder( fix.volume() / 250_ml * cost_scaling, 1.0f ) );
 
-    std::function<bool( const item & )> filter;
```
Comment (kevingranade) [inline src/iuse_actor.cpp]: Merge error?

**Comment by davidpwbrown** [comment]
Yes, no idea how these unrelated changes came in, may make a new PR with cherry-picked commits.

**Comment by ZhilkinSerg** [comment]
Marking as stalled.
